securedrop/journalist_app/sessions.py

```diff
@@ -0,0 +1,278 @@
+            self.uid: Optional[int] = None
+            self.user = None
+        CallbackDict.__init__(self, initial, on_update)
+        self.sid = sid
+        self.token: str = token
+        self.lifetime = lifetime
+        self.is_api = False
+        self.to_destroy = False
+        self.to_regenerate = False
+        self.modified = False
+        self.flash: Optional[Tuple[str, str]] = None
+        self.locale: Optional[str] = None
+
+    def get_token(self) -> Optional[str]:
+        return self.token
+
+    def get_lifetime(self) -> datetime:
+        return datetime.now(timezone.utc) + timedelta(seconds=self.lifetime)
+
+    def set_user(self) -> None:
+        if self.uid is not None:
+            self.user = Journalist.query.get(self.uid)
+        if self.user is None:
+            # The uid has no match in the database, and this should really not happen
+            self.uid = None
+            self.to_destroy = True
+
+    def get_user(self) -> Optional[Journalist]:
+        return self.user
+
+    def get_uid(self) -> Optional[int]:
+        return self.uid
+
+    def set_uid(self, uid: int) -> None:
+        self.user = None
+        self.uid = uid
+
+    def logged_in(self) -> bool:
+        if self.uid is not None:
+            return True
+        else:
+            return False
+
+    def destroy(
+        self,
+        flash: Optional[Tuple[str, str]] = None,
+        locale: Optional[str] = None
+    ) -> None:
+        # The parameters are needed to pass the information to the new session
+        self.locale = locale
+        self.flash = flash
+        self.uid = None
+        self.user = None
+        self.to_destroy = True
+
+    def regenerate(self) -> None:
+        self.to_regenerate = True
+
+
+class SessionInterface(FlaskSessionInterface):
+
+    def _generate_sid(self) -> str:
+        return token_urlsafe(32)
+
+    def _get_signer(self, app: Flask) -> URLSafeTimedSerializer:
+        if not app.secret_key:
+            raise RuntimeError('No secret key set')
+        return URLSafeTimedSerializer(app.secret_key,
+                                      salt=self.salt)
+
+    """Uses the Redis key-value store as a session backend.
+
+    :param redis: A ``redis.Redis`` instance.
+    :param key_prefix: A prefix that is added to all Redis store keys.
+    :param salt: Allows to set the signer salt from the calling interface
+    :param header_name: if use_header, set the header name to parse
+    """
+
+    def __init__(self, lifetime: int, renew_count: int, redis: Optional[Redis],
+                 key_prefix: str, salt: str, header_name: str) -> None:
+        self.serializer = session_json_serializer
+        if redis is None:
+            redis = Redis()
+        self.redis = redis
+        self.lifetime = lifetime
+        self.renew_count = renew_count
+        self.key_prefix = key_prefix
+        self.api_key_prefix = 'api_' + key_prefix
+        self.salt = salt
+        self.api_salt = 'api_' + salt
```

Comment:
In this case yes, I would say that we do not want to mix the SecureDrop Client sessions with the Jounalist Interface sessions even tough they originate from the same credentials. If we drop this, then an API token could be set in a cookie and would work there. While there are no obvious security risks that can I image now, this is still something unwanted and generally unexpected in most applications. In the future, it could also be useful to show active sessions in the admin page for instance.